daemon/images/image_import.go

```diff
@@ -124,22 +71,20 @@ func (i *ImageService) ImportImage(ctx context.Context, src string, repository s
 		}},
 	})
 	if err != nil {
-		return err
+		return "", err
 	}
 
 	id, err := i.imageStore.Create(imgConfig)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	// FIXME: connect with commit code and call refstore directly
 	if newRef != nil {
 		if err := i.TagImageWithReference(id, newRef); err != nil {
-			return err
+			return id, err
```

Comment:
Was returning the `id` on a failure intentional here? Are there uses for still consuming the `id` even if this failed?

I see the current code calling this is discarding it;

```go
id, progressErr = ir.backend.ImportImage(ctx, ref, platform, comment, layerReader, r.Form["changes"])

if progressErr == nil {
	output.Write(streamformatter.FormatStatus("", id.String()))
}
```


_IF_ there's uses for this and we keep this return, we should add a comment here, and also as part of the function's documentation to explain in what cases both an ID, and an error are returned.

Reply:
From the perspective of this function, the image has been created and the ID exists, it's just the tag that failed. Since we don't remove the created image it makes some sense to me that we both return an id and error here.

But you're right that the current usage doesn't make any use of it. I'll change that, push a separate commit and we can see if it makes more sense.

Reply:
Yes, I had that scenario in mind indeed; effectively, we're somewhat pretending the `ImportImage()` to be an atomic operation, but it's not fully (image imported, but not tagged). In order to make the `id` still usable for that scenario, it should either return a specific error (so that the caller can handle it differently), or the operations should be split (`import -> tag`).

Mostly wondering what happens if we import successfully, but fail to tag?

- Is the image accessible (by digest?)
- If so, can it be (manually) tagged later (once / if we support `docker image tag <digest> name:tag`)
- Will it be garbage collected if no action takes place after this (and _when_)?

Reply:
Before containerd image service, if the image fails to be tagged then it's still accessible.
This is what I get when I just replace the call to `TagImageWithReference` with `err = errors.New("simulated error")`
```bash
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker import asdf.tar.zst mynewimage
simulated error
sha256:85bfb97409716fd76c06a64ebfabc4de87a62e12b40b36d322543a093910de08
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
<none>       <none>    85bfb9740971   6 seconds ago   22.3MB
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker tag 85bfb9740971 mynewimage
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
mynewimage   latest    85bfb9740971   6 minutes ago   22.3MB
```


With containerd it's not that easy, because tagging = creating an image, and if the image isn't created then there's nothing that references the underlying content so it will be GC'ed. In this case we should just error the whole operation with error.
That also makes it difficult to split the whole operation into import&tag - because the atomicity of import is dependent on the image service implementation.

Reply:
I dropped this, see: https://github.com/moby/moby/pull/44628#issuecomment-1380046654